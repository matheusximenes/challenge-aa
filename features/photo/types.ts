export interface IPhoto {
    id: string,
    url: string,
    filename: string,
    description: string,
    uploadedBy: string,
    createdAt: string,
    updatedAt: string,
    dimensions: IDimensions,
    resolution: IResolution,
    sizeInBytes: number,
    sharedWith: IUser[],
    favorited: boolean
}

export interface IDimensions {
    height: number,
    width: number
}

export interface IResolution {
    height: number,
    width: number
}

export interface IUser {
    id: string,
    name: string,
    avatar: string
}