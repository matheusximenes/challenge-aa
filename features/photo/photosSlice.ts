import { createSlice } from '@reduxjs/toolkit'
import type { PayloadAction } from '@reduxjs/toolkit'
import { IPhoto } from './types'


interface PhotosState {
  photos: IPhoto[],
  selectedPhoto:  null | IPhoto,
  displayFavorite: boolean
}

const initialState: PhotosState = {
  photos: [],
  selectedPhoto: null,
  displayFavorite: false,
}

type IPhotoID = Pick<IPhoto, "id">

export const photoSlice = createSlice({
  name: 'photos',
  initialState,
  reducers: {
    setPhotos: (state, action: PayloadAction<IPhoto[]>) => {
      state.photos = action.payload
    },
    setSelectedPhoto: (state, action: PayloadAction<IPhotoID | null>) => {
      if(action.payload !== null) {
        const index = state.photos.findIndex(p => p.id === action.payload!.id)
        state.selectedPhoto = state.photos[index]
      } else {
        state.selectedPhoto = null
      }
      
    },
    toggleFavorite: (state, action: PayloadAction<IPhotoID>) => {
      const index = state.photos.findIndex(p => p.id === action.payload.id)
      state.photos[index].favorited = !state.photos[index].favorited;
      state.selectedPhoto = state.photos[index]
    },
    deletePhoto: (state, action: PayloadAction<IPhotoID>) => {
      const newPhotos = state.photos.filter(p => p.id !== action.payload.id)
      state.photos = newPhotos;
      state.selectedPhoto = null
    },
    displayFavorite: (state, action: PayloadAction<boolean>) => {
      state.displayFavorite = action.payload
    }
  },
})

export const { setPhotos, setSelectedPhoto, toggleFavorite, deletePhoto, displayFavorite } = photoSlice.actions

export default photoSlice.reducer