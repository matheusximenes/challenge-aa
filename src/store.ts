import { configureStore } from '@reduxjs/toolkit'
import photosReducer from './features/photo/photosSlice';

export const store = configureStore({
  reducer: {
    photos: photosReducer,
  },
  // middleware: (getDefaultMiddleware) => getDefaultMiddleware({serializableCheck: false}),
})

export type RootState = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch