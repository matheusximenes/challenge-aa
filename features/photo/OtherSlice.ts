import { createSelector, createSlice } from '@reduxjs/toolkit'
import type { PayloadAction } from '@reduxjs/toolkit'
import { IPhoto } from './types'
import { RootState } from '../../src/store'


interface PhotosState {
  photos: IPhoto[],
  selectedPhoto:  null | IPhoto,
  displayFavorite: boolean
}

const initialState: PhotosState = {
  photos: [],
  selectedPhoto: null,
  displayFavorite: false,
}

type IPhotoID = Pick<IPhoto, "id">

export const photoSlice = createSlice({
  name: 'photos',
  initialState,
  reducers: {
    setPhotos: (state, action: PayloadAction<IPhoto[]>) => {
      return {
        ...state,
        photos: action.payload
      }
    },
    setSelectedPhoto: (state, action: PayloadAction<IPhotoID | null>) => {
      if(action.payload !== null) {
        const imgs = state.photos
        const index = imgs.findIndex(p => p.id === action.payload!.id)
        return {
          ...state,
          selectedPhoto: imgs[index]
        }
      } else {
        return {
          ...state,
          selectedPhoto: null
        }
      }
      
    },
    toggleFavorite: (state, action: PayloadAction<IPhotoID>) => {
      const index = state.photos.findIndex(p => p.id === action.payload.id)
      state.photos[index].favorited = !state.photos[index].favorited;
      return {
        ...state,
        selectedPhoto: state.photos[index]
      }
    },
    deletePhoto: (state, action: PayloadAction<IPhotoID>) => {
      const newPhotos = state.photos.filter(p => p.id !== action.payload.id)
      state.photos = newPhotos;
      state.selectedPhoto = null
    },
    displayFavorite: (state, action: PayloadAction<boolean>) => {
      state.displayFavorite = action.payload
    }
  },
})

export const { setSelectedPhoto, toggleFavorite, displayFavorite } = photoSlice.actions
export const state = createSelector((state: RootState) => state, (state: RootState) => state.photos)
export default photoSlice.reducer