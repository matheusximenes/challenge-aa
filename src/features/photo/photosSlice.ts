import { createSelector, createSlice } from '@reduxjs/toolkit'
import type { PayloadAction } from '@reduxjs/toolkit'
import { IPhoto } from './types'
import { RootState } from "../../store"


interface PhotosState {
  photos: IPhoto[],
  selectedPhoto:  null | IPhoto,
  displayFavorite: boolean
}

const initialState: PhotosState = {
  photos: [],
  selectedPhoto: null,
  displayFavorite: false,
}

type IPhotoID = Pick<IPhoto, "id">

export const photoSlice = createSlice({
  name: 'photos',
  initialState,
  reducers: {
    setPhotos(state, action: PayloadAction<IPhoto[]>) {
      state.photos = action.payload
    },
    setSelectedPhoto(state, action: PayloadAction<IPhotoID | null>){
      if(action.payload !== null) {
        const imgs = state.photos
        const index = imgs.findIndex(p => p.id === action.payload!.id)
        state.selectedPhoto = imgs[index]
      } else {
        state.selectedPhoto = null
      }
      
    },
    toggleFavorite(state, action: PayloadAction<IPhotoID>) {
      const index = state.photos.findIndex(p => p.id === action.payload.id)
      state.photos[index].favorited = !state.photos[index].favorited;
    },
    deletePhoto(state, action: PayloadAction<IPhotoID>) {
      state.photos = state.photos.filter(p => p.id !== action.payload.id)
      state.selectedPhoto = null
    },
    displayFavorite(state, action: PayloadAction<boolean>) {
      state.displayFavorite = action.payload
    }
  },
})

export const { setPhotos, setSelectedPhoto, toggleFavorite, deletePhoto, displayFavorite } = photoSlice.actions
export const statePhotos = createSelector((state: RootState) => state.photos, photos => photos)
export default photoSlice.reducer