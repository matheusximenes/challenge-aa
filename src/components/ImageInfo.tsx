import React from "react";
import { IPhoto } from "../../features/photo/types";
import { formatBytes } from "../libs/files";
import { useDispatch, useSelector } from "react-redux";
import {
  deletePhoto,
  setSelectedPhoto,
  toggleFavorite,
} from "../../features/photo/photosSlice";

import { RootState } from "../store";
import { formatDate } from "../libs/dates";

import heartOutlined from "../assets/heart_outlione.svg";
import heart from "../assets/heart_filled.svg";
import Image from "../assets/image-solid.svg";
import XMark from "../assets/xmark-solid.svg";

const ImageInfo = () => {
  const { selectedPhoto } = useSelector((state: RootState) => state.photos);
  const dispatch = useDispatch();

  if (!selectedPhoto) {
    return (
      <div className="photo-info--warn">
        <img alt="" src={Image} />
        <span>No Photo Selected!</span>
      </div>
    );
  }

  const handleDeleteImage = () => {
    dispatch(deletePhoto({ id: selectedPhoto.id }));
  };

  const toggleFavoriteImage = () => {
    dispatch(toggleFavorite({ id: selectedPhoto.id }));
  };

  const unselectImage = () => {
    dispatch(setSelectedPhoto(null));
  };

  return (
    <div className="photo-info">
      <button
        className="photo-info__close-btn"
        onClick={unselectImage}
        aria-label={`Close ${selectedPhoto.filename} Image Details`}
      >
        <img alt="Close Icon" src={XMark} />
      </button>
      <div className="photo-info__image">
        <img src={selectedPhoto.url} alt={selectedPhoto.description} />
      </div>
      <div className="photo-info__small-data">
        <div>
          <h2>{selectedPhoto.filename}</h2>
          <span>{formatBytes(selectedPhoto.sizeInBytes)}</span>
        </div>
        <button
          aria-label={
            selectedPhoto.favorited ? `Unfavorite Image` : `Favorite Image`
          }
          className="photo-info__favorite"
          onClick={toggleFavoriteImage}
        >
          <img
            alt="selectedPhoto.favorited ? `favorited Image` : `No favorited Image`"
            src={selectedPhoto.favorited ? heartOutlined : heart}
          />
        </button>
      </div>
      <h3 className="photo-info__heading">Information</h3>
      <ul className="photo-info__heading-list">
        <li>
          <div>Uploaded by</div>
          <span>{selectedPhoto.uploadedBy}</span>
        </li>
        <li>
          <div>Created</div>
          <span>{formatDate(selectedPhoto.createdAt)}</span>
        </li>
        <li>
          <div>Last Modified</div>
          <span>{formatDate(selectedPhoto.updatedAt)}</span>
        </li>
        <li>
          <div>Dimensions</div>
          <span>
            {selectedPhoto.dimensions?.height} X{" "}
            {selectedPhoto.dimensions?.width}
          </span>
        </li>
        <li>
          <div>Resolution</div>
          <span>
            {selectedPhoto.resolution?.height} X{" "}
            {selectedPhoto.resolution?.width}
          </span>
        </li>
      </ul>

      {selectedPhoto.description && (
        <>
          <h3 className="photo-info__heading--description">Description</h3>
          <p className="photo-info__description">{selectedPhoto.description}</p>
        </>
      )}

      <button className="photo-info__delete-btn" onClick={handleDeleteImage}>
        Delete
      </button>
    </div>
  );
};

export default ImageInfo;
