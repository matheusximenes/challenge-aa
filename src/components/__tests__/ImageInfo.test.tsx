import { cleanup, render, screen } from "@testing-library/react";
import renderer from "react-test-renderer";
import { MOCK_PHOTOS_DATA } from "../../features/photo/__tests__/photosSlice.test";
import ImageInfo from "../ImageInfo";
import ReduxProvider from "../ReduxProvider";
import { store } from "../../store";
import { setPhotos } from "../../features/photo/photosSlice";

describe("imageInfo Component", () => {
  afterEach(cleanup);

  describe("render initial state", () => {
    it.skip("should display initial state - Empty", () => {
      render(
        <ReduxProvider>
          <ImageInfo />
        </ReduxProvider>
      );
      const comp = screen.getByText(/No Photo Selected!/i);
      expect(comp).toBeInTheDocument();
    });
  });

  describe("render selected state", () => {
    it.skip("should display 'title'", () => {
      store.dispatch(setPhotos(MOCK_PHOTOS_DATA));
      render(
        <ReduxProvider>
          <ImageInfo />
        </ReduxProvider>
      );
      const comp = screen.getByRole("heading", {
        name: MOCK_PHOTOS_DATA[0].filename,
        level: 2,
      });
      expect(comp).toBeInTheDocument();
    });
    it.skip("should display 'sizeInBytes'", () => {
      store.dispatch(setPhotos(MOCK_PHOTOS_DATA));
      render(
        <ReduxProvider>
          <ImageInfo />
        </ReduxProvider>
      );
      const comp = screen.getByText(MOCK_PHOTOS_DATA[0].sizeInBytes);
      expect(comp).toBeInTheDocument();
    });

    it("should display 'uploadedBy'", () => {
      store.dispatch(setPhotos(MOCK_PHOTOS_DATA));
      render(
        <ReduxProvider>
          <ImageInfo />
        </ReduxProvider>
      );
      const comp = screen.getByText(MOCK_PHOTOS_DATA[0].uploadedBy);
      expect(comp).toBeInTheDocument();
    });
  });

  describe("Snapshots", () => {
    it("empty state", () => {
      const tree = renderer
        .create(
          <ReduxProvider>
            <ImageInfo />
          </ReduxProvider>
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
