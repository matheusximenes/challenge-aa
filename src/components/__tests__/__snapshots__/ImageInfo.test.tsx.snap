// Vitest Snapshot v1

exports[`imageInfo Component > Snapshots > empty state 1`] = `
<aside
  className="main__aside main__aside--active"
>
  <div
    className="photo-info"
  >
    <button
      aria-label="Close tennessee_female_rubber.jpg Image Details"
      className="photo-info__close-btn"
      onClick={[Function]}
    >
      <img
        alt="Close Icon"
        src="/src/assets/iconography/xmark-solid.svg"
      />
    </button>
    <div
      className="photo-info__image"
    >
      <img
        alt="Laboriosam eligendi inventore officia nemo. Quisquam explicabo voluptatem. Illo laborum facilis."
        src="https://agencyanalytics-api.vercel.app/images/0.jpg"
      />
    </div>
    <div
      className="photo-info__small-data"
    >
      <div>
        <h2>
          tennessee_female_rubber.jpg
        </h2>
        <span>
          4.59 MB
        </span>
      </div>
      <button
        aria-label="Unfavorite Image"
        className="photo-info__favorite"
        onClick={[Function]}
      >
        <img
          alt="selectedPhoto.favorited ? \`favorited Image\` : \`No favorited Image\`"
          src="/src/assets/iconography/heart_outlione.svg"
        />
      </button>
    </div>
    <h3
      className="photo-info__heading"
    >
      Information
    </h3>
    <ul
      className="photo-info__heading-list"
    >
      <li>
        <div>
          Uploaded by
        </div>
        <span>
          Ms. Jimmie Cole
        </span>
      </li>
      <li>
        <div>
          Created
        </div>
        <span>
          July 15, 2017
        </span>
      </li>
      <li>
        <div>
          Last Modified
        </div>
        <span>
          December 16, 2022
        </span>
      </li>
      <li>
        <div>
          Dimensions
        </div>
        <span>
          4800
           X
           
          3200
        </span>
      </li>
      <li>
        <div>
          Resolution
        </div>
        <span>
          72
           X
           
          72
        </span>
      </li>
    </ul>
    <h3
      className="photo-info__heading--description"
    >
      Description
    </h3>
    <p
      className="photo-info__description"
    >
      Laboriosam eligendi inventore officia nemo. Quisquam explicabo voluptatem. Illo laborum facilis.
    </p>
    <button
      className="photo-info__delete-btn"
      onClick={[Function]}
    >
      Delete
    </button>
  </div>
</aside>
`;
